### PR TITLE
Add ADFS support for ClientSecretCredential

### LINKED ISSUE #345: Add ADFS / Azure Stack Support
- On-prem active directory instance support
- Requested by CLI / PSH for .NET and Python

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #345.

### CODE REVIEW

File: sdk/identity/azure-identity/src/client_secret_credential.cpp [hunk 1]
```diff
@@ -27,24 +61,21 @@ Azure::Core::Credentials::AccessToken ClientSecretCredential::GetToken(
...
 
-    auto const& scopes = tokenRequestContext.Scopes;
-    if (!scopes.empty())
     {
-      auto scopesIter = scopes.begin();
-      body << "&scope=" << Url::Encode(*scopesIter);
-
-      auto const scopesEnd = scopes.end();
-      for (++scopesIter; scopesIter != scopesEnd; ++scopesIter)
+      auto const& scopes = tokenRequestContext.Scopes;
```
Comment (RickWinter) [inline sdk/identity/azure-identity/src/client_secret_credential.cpp:75]: Formatting looks off here, Might be github review though.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp [hunk 2]
```diff
@@ -9,6 +9,40 @@
 #include <chrono>
 #include <sstream>
 
+namespace {
```
Comment (RickWinter) [inline sdk/identity/azure-identity/src/client_secret_credential.cpp:12]: This isn't wrong to use the unnamed namespace though I'm not sure it has much advantage over using `Azure::Identity::_detail` within the C++ file.  Afterall the unnamed ends up in `::<unique>::....`.  
From what I understand there is some compiler advantage at link time because no symbol resolution occurs. 
 I'm not sure that is worth it over consistency of `_detail` namespace.

Feel free to leave this as its an implementation detail that we could adjust in the future.

**Review by RickWinter — DISMISSED** [review]
This needs test cases added.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp (hunk 2 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:12]: We should be using unnamed namespaces, unless we need named ones. They were designed for this purpose.
Using unnamed namespaces also allows compiler to make all sorts of right assumptions and make optimizations.
Plus, things from the unnamed namespace won't be made visible by linker.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp (hunk 1 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:75]: If formatting was off, CI wouldfail. Here, indentation is correct. I added extra `{}` to localize `scopes`.

Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:75]: Test cases will be added by #1596.
Right now, there is no infrastructure to mock web responses. The effort to add one testcase would result in more files changing, after which I will be asked to split this PR into multiple parts - this original PR, and the other fixing just the #1596.

**Review by schaabs — APPROVED** [review]

File: sdk/identity/azure-identity/src/client_secret_credential.cpp (hunk 1 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:75]: We seem to be accruing a lot of test debt against that issue. Either prioritize that issue and get the test structure in place, first, OR for new code being checked-in add some unit tests. We need to ensure the code works as intended.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp [hunk 3]
```diff
@@ -9,6 +9,40 @@
...
+// Assumes !scopes.empty()
```
Comment (bhardwajs) [inline sdk/identity/azure-identity/src/client_secret_credential.cpp:13]: Since this assumes `!scopes.empty()`, should there be an early return?

File: sdk/identity/azure-identity/src/client_secret_credential.cpp (hunk 1 above)
Comment (vhvb1989) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:75]: Yeah, at least the test happy path should be there so we know what we add do what we want

File: sdk/identity/azure-identity/src/client_secret_credential.cpp (hunk 3 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:13]: I'd rather not - or otherwise, we need to return an empty string, so the external check for `!scopes.empty()` becomes `!scopesString.empty()`, and that's extra empty work the program will do.

File: sdk/core/azure-core/inc/azure/core/internal/system_clock.hpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+#include <atomic>
+#include <chrono>
+#include <functional>
+
+#include "azure/core/dll_import_export.hpp"
+#include "azure/core/http/transport.hpp"
+
+namespace Azure { namespace Core { namespace _internal {
+  class SystemClock {
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/internal/system_clock.hpp:14]: This is a lot of complexity to mock the system clock.  Can you describe the scenario where we want to replace the system clock on the fly.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/internal/system_clock.hpp:14]: How else would I have an expectation for token expiration value?

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  result.Response = credential.GetToken(tokenRequestContext, Azure::Core::Context());
+
+  return result;
+}
+} // namespace
+
+TEST(ClientSecretCredential, Regular)
+{
+  ClientSecretCredentialOptions options;
+  options.AuthorityHost = "https://authority.url/";
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:78]: This is new. Where can I get the list?

File: sdk/core/azure-core/CMakeLists.txt [hunk 6]
```diff
@@ -56,10 +56,12 @@ set(
     inc/azure/core/internal/diagnostics/log.hpp
     inc/azure/core/internal/hkeyholder.hpp
     inc/azure/core/internal/http/pipeline.hpp
+    inc/azure/core/internal/http/test_transport.hpp
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/core/azure-core/CMakeLists.txt:59]: This is internal. I will be using it from many tests. How else would I mock the HTTP response without huge code repetition. This is a convenient and simple class.

File: sdk/core/azure-core/inc/azure/core/internal/system_clock.hpp (hunk 4 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/internal/system_clock.hpp:14]: Going forward, I'll use this to test retry policy, and token refresh (BearerTokenAuthenticationPolicy)

File: sdk/core/azure-core/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (RickWinter) [reply sdk/core/azure-core/CMakeLists.txt:59]: Its fine that its internal and used for tests.  It should reside in the test directory.  See TransportAdapter for an example on to replace the replace transport adapter in tests.

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (RickWinter) [reply sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:78]: I'll see if I can dig up the doc.  The gist is you don't want to put a url to something we don't own.  By using a controlled domain (or nothing at all "") you prevent inadvertent information disclosures.

File: sdk/core/azure-core/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/core/azure-core/CMakeLists.txt:59]: I can't have it in test directory without repetition, because it lives in Core, and I consume it from Identity. We don't have a mechanism to include test headers from other library.

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:78]: In this test, this is the part of what I am testing, I definitely don't want to use "". If you can't find it, would `microsoft.com` and `azure.com` be fine?

File: sdk/core/azure-core/inc/azure/core/internal/system_clock.hpp (hunk 4 above)
Comment (RickWinter) [reply sdk/core/azure-core/inc/azure/core/internal/system_clock.hpp:14]: It would be simpler to back date the expiration time on the token rather than artificially moving the system clock forward/backward.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/internal/system_clock.hpp:14]: Token does not have the expiration date, it has "expires_in", which is in seconds from now. And we can't be precise about "now". Even if I get timestamp for before and after, and verify that the expiration falls between them, very soon I'll need to write a test for BererTokenAuthenticationPolicy and RetryPolicy, where we'll need to inject the values for DateTime.Now + Thread.Sleep.

File: sdk/core/azure-core/src/datetime.cpp [hunk 7]
```diff
@@ -40,6 +40,20 @@ DateTime GetSystemClockEpoch()
...
+                                  std::chrono::system_clock::time_point::max().time_since_epoch())
+                                  .count();
+
+  auto const systemClockEpoch = GetSystemClockEpoch().time_since_epoch().count();
+
+  constexpr auto repMax = std::numeric_limits<DateTime::clock::duration::rep>::max();
+
+  return DateTime(DateTime::time_point(
+      DateTime::duration(systemClockMax + std::min(systemClockEpoch, (repMax - systemClockMax)))));
+}
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/datetime.cpp:55]: What's the goal of this method and why do we need it? Please add some comments. How' it different from `std::chrono::system_clock::time_point::max()`?

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp [hunk 8]
```diff
@@ -25,7 +25,7 @@ namespace Azure { namespace Identity {
   /**
    * @brief Defines options for token authentication.
    */
-  struct ClientSecretCredentialOptions : public Azure::Core::_internal::ClientOptions
+  struct ClientSecretCredentialOptions : public Azure::Core::Credentials::TokenCredentialOptions
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:28]: Why was this change required instead of deriving directly from `ClientOptions`?

File: sdk/core/azure-core/inc/azure/core/credentials/token_credential_options.hpp [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+#pragma once
+
+#include "azure/core/internal/client_options.hpp"
+
+namespace Azure { namespace Core { namespace Credentials {
+  /**
+   * @brief Defines options for #Azure::Core::Credentials::TokenCredential.
+   */
+  struct TokenCredentialOptions : public Azure::Core::_internal::ClientOptions
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/credentials/token_credential_options.hpp:17]: Needs CL entry.

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp [hunk 10]
```diff
@@ -64,12 +64,32 @@ namespace Azure { namespace Identity {
...
         std::string clientSecret,
-        ClientSecretCredentialOptions options = ClientSecretCredentialOptions())
+        ClientSecretCredentialOptions options)
         : m_tenantId(std::move(tenantId)), m_clientId(std::move(clientId)),
           m_clientSecret(std::move(clientSecret)), m_options(std::move(options))
     {
     }
 
+    /**
+     * @brief Construct a Client Secret credential.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:74]: This is an odd casing choice. Was it intentional?

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp [hunk 11]
```diff
@@ -64,12 +64,32 @@ namespace Azure { namespace Identity {
...
+    explicit ClientSecretCredential(
+        std::string tenantId,
+        std::string clientId,
+        std::string clientSecret,
+        Azure::Core::Credentials::TokenCredentialOptions const& options
+        = Azure::Core::Credentials::TokenCredentialOptions())
+        : m_tenantId(std::move(tenantId)), m_clientId(std::move(clientId)),
+          m_clientSecret(std::move(clientSecret))
+    {
+      static_cast<Azure::Core::Credentials::TokenCredentialOptions&>(m_options) = options;
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:90]: nit: Add a comment explaining why this cast is OK.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp [hunk 12]
```diff
@@ -9,6 +9,40 @@
...
+std::string FormatScopes(std::vector<std::string> const& scopes, bool asResource)
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/client_secret_credential.cpp:14]: What does `asResource` mean?

File: sdk/identity/azure-identity/src/client_secret_credential.cpp [hunk 13]
```diff
@@ -9,6 +9,40 @@
...
+{
+  if (asResource && scopes.size() == 1)
+  {
+    auto resource = scopes[0];
+    constexpr char suffix[] = "/.default";
+    constexpr int suffixLen = sizeof(suffix) - 1;
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/client_secret_credential.cpp:20]: Why are we doing `-1`. Add a comment.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp [hunk 14]
```diff
@@ -9,6 +9,40 @@
...
+        && resource.find(suffix, resourceLen - suffixLen) != std::string::npos)
+    {
+      resource = resource.substr(0, resourceLen - suffixLen);
+    }
+
+    return Azure::Core::Url::Encode(resource);
+  }
+
+  auto scopesIter = scopes.begin();
+  auto scopesStr = Azure::Core::Url::Encode(*scopesIter);
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/client_secret_credential.cpp:34]: nit: re-add the comment here, that we assume `scopes.size() > 0`

File: sdk/identity/azure-identity/src/environment_credential.cpp [hunk 15]
```diff
@@ -53,28 +54,29 @@ EnvironmentCredential::EnvironmentCredential()
     {
       if (authority != nullptr)
       {
-        ClientSecretCredentialOptions options;
-        options.AuthorityHost = authority;
+        ClientSecretCredentialOptions clientSecretCredentialOptions;
+        static_cast<Core::_internal::ClientOptions&>(clientSecretCredentialOptions) = options;
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/environment_credential.cpp:58]: Shouldn't this be `TokenCredentialOptions`?
```suggestion
        static_cast<Core::Credentials::TokenCredentialOptions&>(clientSecretCredentialOptions) = options;
```

What's this line doing? Is it only copying over the fields into the local `clientSecretCredentialOptions` that come from the base type `options`? Does it leave the other fields alone?

Also, add comment why static cast is OK here.

File: sdk/identity/azure-identity/src/environment_credential.cpp [hunk 16]
```diff
@@ -53,28 +54,29 @@ EnvironmentCredential::EnvironmentCredential()
...
+            tenantId, clientId, clientSecret, clientSecretCredentialOptions));
       }
       else
       {
-        m_credentialImpl.reset(new ClientSecretCredential(tenantId, clientId, clientSecret));
+        m_credentialImpl.reset(
+            new ClientSecretCredential(tenantId, clientId, clientSecret, options));
       }
     }
     // TODO: These credential types are not implemented. Uncomment when implemented.
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/environment_credential.cpp:70]: Since you are hear and modifying this code already, remove this TODOvcomment from source and file an issue.

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+
+  ClientSecretCredential credential(tenantId, clientId, clientSecret, credentialOptions);
+  result.Response.AccessToken = credential.GetToken(tokenRequestContext, Azure::Core::Context());
+  result.Response.Latest = std::chrono::system_clock::now();
+
+  return result;
+}
+} // namespace
+
+TEST(ClientSecretCredential, Regular)
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:69]: How can I see the test coverage for this change and azure-identity overall?

It's not visible here: https://dev.azure.com/azure-sdk/public/_build/results?buildId=808213&view=results

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+  options.AuthorityHost = "https://microsoft.com/";
+  auto const actual = TestClientSecretCredential(
+      "adfs",
+      "fedcba98-7654-3210-0123-456789abcdef",
+      "CLIENTSECRET",
+      options,
+      {{"https://azure.com/.default"}},
+      "{\"expires_in\":3600, \"access_token\":\"ACCESSTOKEN1\"}");
+
+  EXPECT_EQ(actual.Request.AbsoluteUrl, "https://microsoft.com/adfs/oauth2/token");
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:120]: Is this the only difference between the two tests `AzureStack` and `Regular`?

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+    EXPECT_NE(actual.Request.Headers.find("Content-Length"), actual.Request.Headers.end());
+    EXPECT_EQ(
+        actual.Request.Headers.at("Content-Length"), std::to_string(sizeof(expectedBody) - 1));
+  }
+
+  EXPECT_NE(actual.Request.Headers.find("Content-Type"), actual.Request.Headers.end());
+  EXPECT_EQ(actual.Request.Headers.at("Content-Type"), "application/x-www-form-urlencoded");
+
+  EXPECT_EQ(actual.Response.AccessToken.Token, "ACCESSTOKEN1");
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:101]: Also look for host here, similar to below?
```C++
  EXPECT_NE(actual.Request.Headers.find("Host"), actual.Request.Headers.end());
  EXPECT_EQ(actual.Request.Headers.at("Host"), "microsoft.com");
```

File: sdk/identity/azure-identity/test/ut/environment_credential.cpp [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+
+    try
+    {
+      Environment::SetVariables(NewEnv);
+    }
+    catch (...)
+    {
+      Environment::SetVariables(m_originalEnv);
+      throw;
+    }
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/environment_credential.cpp:111]: What are we trying to do here?

File: sdk/identity/azure-identity/test/ut/environment_credential.cpp [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+      std::string const& clientCertificatePath)
+  {
+    std::map<std::string, std::string> const NewEnv = {
+        {"AZURE_TENANT_ID", tenantId},
+        {"AZURE_CLIENT_ID", clientId},
+        {"AZURE_CLIENT_SECRET", clientSecret},
+        {"AZURE_AUTHORITY_HOST", authorityHost},
+        {"AZURE_USERNAME", username},
+        {"AZURE_PASSWORD", password},
+        {"AZURE_CLIENT_CERTIFICATE_PATH", clientCertificatePath},
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/environment_credential.cpp:95]: Correct me if I am mistaken, but this test seems to be overriding some of the commonly used env vars on any contributor developer machine. Is there a way to get the test coverage but use custom env var names rather than things like `AZURE_TENANT_ID`?

I would prefer if running unit tests didn't change my dev machine's env vars.

Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/environment_credential.cpp:95]: These changes are not machine-wide. They are for the current process only.

File: sdk/identity/azure-identity/test/ut/environment_credential.cpp (hunk 20 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/environment_credential.cpp:111]: If an exception occurs in constructor, we restore environment variables.

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp (hunk 19 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:101]: We do not set Host header for non-ADFS requests.

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp (hunk 18 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:120]: Not the only - you can see by the changes to the ClientSecretCredential - it is the Host header, and different auth scope.

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp (hunk 8 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:28]: It would look strange if there was CredentialOptions, and ClientSecretCredentialOptions wouldn't derive from it. Plus, other SDKs have it.

File: sdk/core/azure-core/inc/azure/core/credentials/token_credential_options.hpp (hunk 9 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/credentials/token_credential_options.hpp:17]: It is not a breaking change, and has the default value. All the code you could compile before can still compile now.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp (hunk 12 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:14]: it is internal function. You can only have single scope as resource, and when you want to build resource, you don't add scope suffix.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp (hunk 13 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:20]: I don't think this necessarily requires a comment, this is the part of the language - when an array is initialized from a literal, `\0` is added to the array - https://en.cppreference.com/w/c/language/array_initialization.

Comment (ahsonkhan) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:20]: > I don't think this necessarily requires a comment

Source comments are rarely **necessary**. It's just meant for helpful hints for future maintainability when code changes, moves around. What happens if I change line 19 tomorrow to `std::string`, now line 20 is broken, if I don't update `-1`, correct?

It's up to your discretion, as the PR author, what's clear/obvious by just reading the code.

File: sdk/identity/azure-identity/src/environment_credential.cpp (hunk 15 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/src/environment_credential.cpp:58]: @antkmsft does this need to be fixed for corectness?

File: sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp (hunk 8 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/inc/azure/identity/client_secret_credential.hpp:28]: Gotcha, makes sense.

File: sdk/identity/azure-identity/test/ut/environment_credential.cpp (hunk 20 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/ut/environment_credential.cpp:111]: Why? Why do we need to be resilient against exceptions to restore env vars in the test?

File: sdk/identity/azure-identity/src/environment_credential.cpp (hunk 15 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/environment_credential.cpp:58]: For now we are fine, but yes, it needs fixing - I'll grab this plus a few other comments and create a follow-up PR.

File: sdk/identity/azure-identity/test/ut/client_secret_credential.cpp (hunk 17 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/client_secret_credential.cpp:69]: I don't know how the code coverage integration works, but if you don't see it, it most likely was not enable for identity. We need a work item to enable it.

File: sdk/identity/azure-identity/test/ut/environment_credential.cpp (hunk 20 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/test/ut/environment_credential.cpp:111]: We don't have to, just a habit. Still, it may result in less noisy tests when something fails, when we restore it to the way it was before we failed.

File: sdk/identity/azure-identity/src/client_secret_credential.cpp (hunk 13 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/src/client_secret_credential.cpp:20]: I can add it, but if we change it to `std::string`, `suffixLen` should be removed entirely, it is no longer necessary. So, the only context in which it makes sense, is when `suffix` is a C-string, so it is 0-terminated, so that "- 1" is needed.

But even without the "-1", `sizeof(std::string())` is broken. It will be the size of the `std::string` class, not the length of the string. `length()` should be used.

File: sdk/core/azure-core/CMakeLists.txt [hunk 22]
```diff
@@ -46,6 +46,7 @@ set(
     ${CURL_TRANSPORT_ADAPTER_INC}
     ${WIN_TRANSPORT_ADAPTER_INC}
     inc/azure/core/credentials/credentials.hpp
+    inc/azure/core/credentials/token_credential_options.hpp
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CMakeLists.txt:49]: We need to include these headers to `core.hpp` and add tests to the simplified header.
Fixed in https://github.com/Azure/azure-sdk-for-cpp/pull/2070
